src/main/java/com/minecolonies/entity/ai/citizen/lumberjack/Tree.java

```diff
@@ -80,128 +97,127 @@ private Tree()
     {
         Block block = BlockPosUtil.getBlock(world, log);
         if (block.isWood(world, log))
         {
-            variant = world.getBlockState(log).getValue(BlockNewLog.VARIANT);
-            location = getBaseLog(world, log);
+            if (block instanceof BlockOldLog)
+            {
+                variant = world.getBlockState(log).getValue(BlockOldLog.VARIANT);
+            }
+            else if (block instanceof BlockNewLog)
+            {
+                variant = world.getBlockState(log).getValue(BlockNewLog.VARIANT);
+            }
+            else
+            {
+                variant = BlockPlanks.EnumType.OAK;
+            }
+
             woodBlocks = new LinkedList<>();
-            checkTree(world, getTopLog(world, log));
+            location = log;
+            topLog = log;
+
+            addAndSearch(world, log);
+
+            checkTree(world, topLog);
             stumpLocations = new ArrayList<>();
+            woodBlocks.clear();
         }
     }
 
     /**
-     * Get's the base log of the tree
+     * For use in PathJobFindTree.
      *
-     * @param world The entity world
-     * @param pos   The coordinates
-     * @return the base log position
+     * @param world the world.
+     * @param pos   The coordinates.
+     * @return true if the log is part of a tree.
      */
-    private static BlockPos getBaseLog(@NotNull IBlockAccess world, BlockPos pos)
+    public static boolean checkTree(@NotNull IBlockAccess world, BlockPos pos)
     {
-        BlockPos basePos = pos;
-        while (world.getBlockState(basePos.down()).getBlock().isWood(world, basePos))
+        //Is the first block a log?
+        if (!world.getBlockState(pos).getBlock().isWood(world, pos))
         {
-            basePos = basePos.down();
+            return false;
         }
-        return basePos;
+
+        final Tuple<BlockPos, BlockPos> baseAndTOp = getBottomAndTopLog(world, pos, new LinkedList<>(), null, null);
+
+        //Get base log, should already be base log.
+        final BlockPos basePos = baseAndTOp.getFirst();
+
+        //Make sure tree is on solid ground and tree is not build above cobblestone.
+        return world.getBlockState(basePos.down()).getMaterial().isSolid()
+                 && world.getBlockState(basePos.down()).getBlock() != Blocks.COBBLESTONE
+                 && hasEnoughLeaves(world, baseAndTOp.getSecond());
     }
 
     /**
-     * Checks if the found log is part of a tree
+     * Adds a log and searches for further logs(Breadth first search).
      *
-     * @param world  The world the tree is in
-     * @param topLog The most upper log of the tree
+     * @param world The world the log is in.
+     * @param log   the log to add.
+     * @return a tuple containing, first: bottom log and second: top log.
      */
-    private void checkTree(@NotNull World world, @NotNull BlockPos topLog)
+    @NotNull
+    private static Tuple<BlockPos, BlockPos> getBottomAndTopLog(
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "getBottomAndTopLog" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)